### PR TITLE
feat: set lower and upper bin levels for normalise

### PR DESCRIPTION
**Pull request by LachlanNewman**
See https://github.com/lovell/sharp/issues/3574#issuecomment-1449124779

### CODE REVIEW

**Comment by lovell** [comment]
Hi, thanks for the PR, did you need this change to solve your original problem?

**Comment by LachlanNewman** [comment]
> Hi, thanks for the PR, did you need this change to solve your original problem?

Hey @lovell which problem are you referring to ?

**Comment by jonseppanen** [comment]
This in general is a very welcomed/needed addition to the library. Being able to adjust images with weird levels (e.g all at one end or the other) is sorely needed.

**Comment by lovell** [comment]
@LachlanNewman I was referring to #3574, which I think you've been able to solve by increasing brightness. I was wondering if the change in this PR also allowed you to solve the same problem without the need to alter brightness?

File: lib/operation.js [hunk 1]
```diff
@@ -474,8 +474,10 @@ function negate (options) {
  * @param {Boolean} [normalise=true]
  * @returns {Sharp}
  */
-function normalise (normalise) {
+function normalise (normalise, lowerBin, upperBin) {
```
Comment (lovell) [inline lib/operation.js:477]: How about having the API look a bit more like the following?
```js
.normalise({ lower: 1, upper: 99 })
```
(I'm happy to drop the `normalise` as-a-Boolean parameter as it's a little confusing.)

File: lib/operation.js [hunk 2]
```diff
@@ -474,8 +474,10 @@ function negate (options) {
...
   this.options.normalise = is.bool(normalise) ? normalise : true;
+  this.options.normaliseLowerBin = is.number(lowerBin) ? lowerBin : 1;
```
Comment (lovell) [inline lib/operation.js:479]: Could validate these via `is.inRange()`.

File: src/operations.cc [hunk 3]
```diff
@@ -68,8 +67,8 @@ namespace sharp {
     // Extract luminance
     VImage luminance = lab[0];
     // Find luminance range
-    int const min = luminance.percent(1);
-    int const max = luminance.percent(99);
+    int const min = luminance.percent(lowerBin);
```
Comment (lovell) [inline src/operations.cc:70]: Could optimise slightly here when the values are 0 or 100, which would be `min()` or `max()` respectively.

File: test/unit/normalize.js [hunk 4]
```diff
@@ -20,10 +20,8 @@ describe('Normalization', function () {
   it('spreads rgb image values between 0 and 255', function (done) {
     sharp(fixtures.inputJpgWithLowContrast)
       .normalise()
-      .raw()
-      .toBuffer(function (err, data, info) {
+      .toFile('test2.jpg', function (err) {
```
Comment (lovell) [inline test/unit/normalize.js:23]: Was this change intentional?

**Review by lovell — CHANGES_REQUESTED** [review]
Thanks again, I've left a few questions/comments inline.

**Comment by LachlanNewman** [comment]
> @LachlanNewman I was referring to #3574, which I think you've been able to solve by increasing brightness. I was wondering if the change in this PR also allowed you to solve the same problem without the need to alter brightness?

Nope, #3574 was for a specific image with a very low range of pixels. This one is to allow normalizing the images based on a particular range of pixels.

File: test/unit/normalize.js (hunk 4 above)
Comment (LachlanNewman) [reply test/unit/normalize.js:23]: no this was an accident i have fixed it.

File: lib/operation.js (hunk 1 above)
Comment (LachlanNewman) [reply lib/operation.js]: [resolved](https://github.com/lovell/sharp/pull/3578/commits/8045250a2697591b40bef01db1c9c067587b1973)

File: lib/operation.js (hunk 2 above)
Comment (LachlanNewman) [reply lib/operation.js:482]: [resolved](https://github.com/lovell/sharp/pull/3578/commits/2b16433d97e4cb95a07369b5e8fa68cc80d9b905)

File: src/operations.cc (hunk 3 above)
Comment (LachlanNewman) [reply src/operations.cc]: [resolved](https://github.com/lovell/sharp/pull/3578/commits/f0cd634c86f5734d2a534b4becc594170e8195df)

**Review by lovell — CHANGES_REQUESTED** [review]
Thank you for the updates, one further comment inline. I'll update the target branch for this PR to `flow` for inclusion in the next release, please can you rebase against that branch. This also means you will need to include a TypeScript definition update in `lib/index.d.ts`.

File: lib/operation.js [hunk 5]
```diff
@@ -474,8 +474,13 @@ function negate (options) {
  * @param {Boolean} [normalise=true]
  * @returns {Sharp}
  */
-function normalise (normalise) {
-  this.options.normalise = is.bool(normalise) ? normalise : true;
+function normalise (lowerBin, upperBin) {
+  if(!is.inRange(lowerBin,0,100 ) || !is.inRange(upperBin,0,100)){
```
[the remark this answers is not in the document]
Comment (LachlanNewman) [reply lib/operation.js]: fixed

**Comment by LachlanNewman** [comment]
@lovell fyi see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64616

**Comment by LachlanNewman** [comment]
@lovell see https://github.com/lovell/sharp/pull/3583

**Comment by lovell** [comment]
Sorry I wasn't clear about the TypeScript definition changes. Now that you've rebased against `flow` you'll hopefully have seen what used to be published as `@types/sharp` is now part of *this* repo in the `lib/index.d.ts` file and will be published as part of `@sharp` directly in the next release - please see #3369. This means any API changes in the next release now need to be included in this repo instead of DefinitelyTyped.
